### PR TITLE
[VCARB-94] - Added new enroll cmd

### PR DESCRIPTION
**Pull request by potofpie**
https://www.loom.com/share/9dc0a348851d4a1f8305b6c9c227e673

Testing:
- run this backend branch locally https://github.com/shopmonkeyus/backend/pull/7018
- Create a server in postman pointed at your local env
- Grab the `code`
- example: go run . enroll --code {code} --api-url https://api-de35bc.shopmonkey.cloud

### CODE REVIEW

File: cmd/enroll.go [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+			os.Exit(1)
+		}
+		defer file.Close()
+
+	},
+}
+
+func init() {
+	rootCmd.AddCommand(enrollCmd)
+	enrollCmd.Flags().String("code", "", "the enrollment code from HQ")
```
Comment (robindiddams) [inline cmd/enroll.go:101]: lets just make it take the code as the first and only arg, use `Args:  cobra.ExactArgs(1),` in your cmd declaration and change `Use: "enroll code"` and use args[0] for the code

File: cmd/enroll.go [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+			os.Exit(1)
+		}
+		if !enrollResp.Success {
+			logger.Error("failed to start enroll: %s", enrollResp.Message)
+		}
+		if _, err := os.Stat("dataDir"); os.IsNotExist(err) {
+			os.Mkdir("dataDir", 0755)
+		}
+
+		tokenFile := "dataDir/token.json"
```
Comment (robindiddams) [inline cmd/enroll.go:83]: this'll break on windows, cant use slash, 
```suggestion
		tokenFile := filepath.Join("dataDir", "token.json")
```

also can you make `dataDir` come in through a cli [arg like server](https://github.com/shopmonkeyus/eds-server/blob/f50587884f30c5642bdd328ec3773aaf4b255f56/cmd/server.go#L836-L840)

File: cmd/enroll.go [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+		if err != nil {
+			logger.Error("error getting api url: %s", err)
+			os.Exit(1)
+		}
+
+		req, err := http.NewRequest("GET", *apiURL+"/v3/eds/internal/enroll/"+code, nil)
+		if err != nil {
+			logger.Error("error creating request: %s", err)
+			os.Exit(1)
+
```
Comment (robindiddams) [inline cmd/enroll.go:57]: whitespace

File: cmd/enroll.go [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+		"P": "https://api.shopmonkey.cloud/",
+		"S": "https://sandbox-api.shopmonkey.cloud/",
+		"E": "https://edge-api.shopmonkey.cloud/",
+		"L": "http://localhost:3101",
+	}
+
+	if url, exists := apiUrls[firstLetter]; exists {
+		return &url, nil
+	}
+	return nil, fmt.Errorf("Invalid first letter")
```
Comment (robindiddams) [inline cmd/enroll.go:34]: no formatting, use errors.New
```suggestion
	return nil, errors.New("Invalid first letter")
```

File: cmd/enroll.go [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+var enrollCmd = &cobra.Command{
+	Use:   "enroll",
+	Short: "Enroll a new server and get api key",
+	Run: func(cmd *cobra.Command, args []string) {
+		logger := newLogger(cmd)
+		logger = logger.WithPrefix("[enroll]")
+		code := mustFlagString(cmd, "code", true)
+
+		firstLetter := code[0:1]
+		apiURL, err := getApiUrl(firstLetter)
```
Comment (robindiddams) [inline cmd/enroll.go:46]: this is kinda silly but can you _also_ support providing a url via api-url flag so we can test this on previews

**Review by robindiddams — CHANGES_REQUESTED** [review]

File: cmd/enroll.go (hunk 5 above)
Comment (potofpie) [reply cmd/enroll.go:46]: Yep! Makes sense.

File: cmd/enroll.go [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+		"P": "https://api.shopmonkey.cloud/",
+		"S": "https://sandbox-api.shopmonkey.cloud/",
+		"E": "https://edge-api.shopmonkey.cloud/",
+		"L": "http://localhost:3101",
+	}
+
+	if url, exists := apiUrls[firstLetter]; exists {
+		return &url, nil
+	}
+	return nil, errors.New("invalid first letter")
```
Comment (robindiddams) [inline cmd/enroll.go:35]: maybe make it a bit more vague
```suggestion
	return nil, errors.New("invalid code")
```

File: cmd/enroll.go [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+	Token string `json:"token"`
+}
+
+type enrollResponse struct {
+	Success bool            `json:"success"`
+	Message string          `json:"message"`
+	Data    enrollTokenData `json:"data"`
+}
+
+func getApiUrl(firstLetter string) (*string, error) {
```
Comment (robindiddams) [inline cmd/enroll.go:24]: Go convention is to use [all caps for acronyms](https://google.github.io/styleguide/go/decisions#initialisms)
```suggestion
func getAPIURL(firstLetter string) (*string, error) {
```

**Review by robindiddams — COMMENTED** [review]
lookin better!

File: cmd/enroll.go [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+	}
+
+	if url, exists := apiUrls[firstLetter]; exists {
+		return &url, nil
+	}
+	return nil, errors.New("invalid first letter")
+}
+
+var enrollCmd = &cobra.Command{
+	Use:   "enroll code",
```
Comment (jhaynie) [inline cmd/enroll.go:40]:
```suggestion
	Use:   "enroll [code]",
```

File: cmd/enroll.go [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+		tokenFile := filepath.Join(dataDir, "token.json")
+		file, err := os.Create(tokenFile)
+		if err != nil {
+			logger.Error("failed to create token file: %w", err)
+			os.Exit(1)
+		}
+		_, err = file.WriteString(enrollResp.Data.Token)
+		if err != nil {
+			logger.Error("failed to write to token file: %w", err)
+			os.Exit(1)
```
Comment (jhaynie) [inline cmd/enroll.go:96]: use Fatal for all of these errors and remove exit

File: cmd/enroll.go [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+		}
+		defer file.Close()
```
Comment (jhaynie) [inline cmd/enroll.go:98]: just remove defer since its the last operation

File: cmd/root.go [hunk 11]
```diff
@@ -173,7 +173,8 @@ func getDataDir(cmd *cobra.Command, logger logger.Logger) string {
 	dataDir, _ = filepath.Abs(filepath.Clean(dataDir))
 
 	if !util.Exists(dataDir) {
-		logger.Fatal("data directory %s does not exist. please create the directory and retry again.", dataDir)
+		os.Mkdir(dataDir, 0755)
```
Comment (jhaynie) [inline cmd/root.go:176]: i think we need to make data dir 0700 given the sensitivity

File: cmd/enroll.go [hunk 12]
```diff
@@ -0,0 +1,112 @@
...
+			}
+			apiURL = *maybeApiURL
+		}
+
+		req, err := http.NewRequest("GET", apiURL+"/v3/eds/internal/enroll/"+code, nil)
+		if err != nil {
+			logger.Error("error creating request: %s", err)
+			os.Exit(1)
+		}
+		resp, err := http.DefaultClient.Do(req)
```
Comment (jhaynie) [inline cmd/enroll.go:66]: see HTTPRetry and use it inside internal package

**Review by jhaynie — CHANGES_REQUESTED** [review]

File: cmd/enroll.go [hunk 13]
```diff
@@ -0,0 +1,112 @@
...
+		tokenFile := filepath.Join(dataDir, "token.json")
+		file, err := os.Create(tokenFile)
+		if err != nil {
+			logger.Fatal("failed to create token file: %w", err)
+		}
+		jsonData, _ := json.Marshal(enrollResp.Data)
+		if err != nil {
+			logger.Fatal("Error converting to JSON: %v", err)
+		}
+		_, err = file.WriteString(string(jsonData))
```
Comment (robindiddams) [inline cmd/enroll.go:95]:
```suggestion
		jsonData, err := json.Marshal(enrollResp.Data)
		if err != nil {
			logger.Fatal("Error converting to JSON: %v", err)
		}
		_, err = file.Write(jsonData)
```

**Review by robindiddams — CHANGES_REQUESTED** [review]
ya so close!

File: cmd/root.go [hunk 14]
```diff
@@ -173,7 +173,8 @@ func getDataDir(cmd *cobra.Command, logger logger.Logger) string {
 	dataDir, _ = filepath.Abs(filepath.Clean(dataDir))
 
 	if !util.Exists(dataDir) {
-		logger.Fatal("data directory %s does not exist. please create the directory and retry again.", dataDir)
+		os.Mkdir(dataDir, 0700)
```
Comment (jhaynie) [inline cmd/root.go:176]: oops... one more ... use os.MkdirAll in case there are multiple directory segments that need to be created.

**Review by jhaynie — CHANGES_REQUESTED** [review]

**Review by jhaynie — APPROVED** [review]
💯

**Pull request merged**